scylla/src/client/session.rs

```diff
@@ -1253,41 +1251,91 @@ impl Session {
+            cluster_state: &ClusterState,
+            on_all_shards: bool,
+        ) -> Result<PreparedStatement, PrepareError> {
+            // Find the first result that is Ok, or Err if all failed.
+            let working_connections = if on_all_shards {
+                itertools::Either::Left(cluster_state.iter_working_connections_to_shards()?)
+            } else {
+                itertools::Either::Right(cluster_state.iter_working_connections_to_nodes()?)
+            };
+
+            let preparations =
+                working_connections.map(|c| async move { c.prepare_raw(statement).await });
+            let raw_prepared_statements_results = join_all(preparations).await;
+
+            let mut raw_prepared_statements_results_iter =
+                raw_prepared_statements_results.into_iter();
+
+            // Safety: We pass a nonempty iterator, so there will be at least one result.
+            let first_ok_or_error = raw_prepared_statements_results_iter
+                .by_ref()
+                .find_or_last(|res| res.is_ok())
+                .unwrap(); // Safety: there is at least one connection.
+
+            match first_ok_or_error {
+                Ok(raw_prepared) => {
+                    // Verify id equality.
+                    if let Some(raw_prepared_with_different_id) =
+                        raw_prepared_statements_results_iter
+                            .flatten()
+                            .find(|other_prepared| {
+                                raw_prepared.prepared_response.id
+                                    != other_prepared.prepared_response.id
+                            })
+                    {
+                        tracing::error!(
+                            "Got differing ids upon statement preparation: statement \"{}\", id1: {:?}, id2: {:?}",
+                            statement.contents,
+                            raw_prepared.prepared_response.id,
+                            raw_prepared_with_different_id.prepared_response.id
+                        );
+
+                        return Err(PrepareError::PreparedStatementIdsMismatch);
+                    }
 
-            // Collect all tracing ids from prepare() queries in the final result
-            prepared.prepare_tracing_ids.extend(statement.tracing_id);
+                    let mut prepared = raw_prepared.into_prepared_statement();
+
+                    // This is the first preparation that succeeded.
+                    // Let's return the PreparedStatement.
+                    prepared.set_partitioner_name(
+                        Session::extract_partitioner_name(&prepared, cluster_state)
+                            .and_then(PartitionerName::from_str)
+                            .unwrap_or_default(),
+                    );
+                    Ok(prepared)
+                }
+                Err(err) => Err(PrepareError::AllAttemptsFailed { first_attempt: err }),
+            }
```

Comment:
I don't see why this is a nested function. Please make it just a private top-level function, like prepare_nongeneric.